Fix limiting step re-execute bug (RIPD-1368)

seelabs opened the pull request:
The deferred credits table can compute a balance that's different from the
ledger balance.

Syntax:
A number written with no decimal means that number exactly. I.e. "12". A number
written with a decimal means that number has a non-zero digit at the lowest
order digit. I.e. "12.XX" means a number like "12.00000000000005"

Consider the following payment:
alice (USD) -> USD/XRP -> (XRP) Bob
Alice initially has 12.XX USD in her account.
The strand is used to debit alice the following amounts:
1) Debit alice 5
2) Debit alice 0.XX
3) Debit alice 3.XX

The next time the strand is explored, alice has a USD/XRP offer on the books,
and her account is credited:

1) Credit alice 20

When the beginning of the strand is reached, consider what happens when alice is
a limiting step. Calculate how much we can get out the step. According to the
deferred credit table this is:
12.XX - (5 + 0.XX + 3.XX)

This is also limited by alice's balance, which is large thanks to the credit she
received in the book step.

Now that the step has calculated how much we can get out, throw out the
sandbox (the one with the credit), and re-execute. However, the following error
occurs. We asked for 12.XX - (5 + 0.XX + 3.XX). However, the ledger has
calculated that alice has:
((12.XX - 5) - 0.XX) - 3.XX

That's a problem, because that number is smaller. Notice that there are two
precision losing operations in the deferred credits table:
1) The 5 + 0.XX step
2) The 12.XX - (total of debits). (Notice total of debits is < 10)

However, there is only one precision losing operation in the ledger calculation:
1) (Subtotal of 12.XX-5) - 0.XX

That means the calculation for the ledger results in a number that's smaller
than the deferred credits. Flow detects this as a re-execution error.

@scottschurr @nbougalis

scottschurr commented on src/ripple/ledger/impl/PaymentSandbox.cpp:
```diff
@@ -186,15 +186,26 @@ PaymentSandbox::balanceHook (AccountID const& account,
             if (auto adj = curSB->tab_.adjustments (account, issuer, currency))
             {
                 delta += adj->debits;
                 lastBal = adj->origBalance;
```
Not a correction, just a question for my own understanding.

I'm assuming that several of the recursively referenced `Sandboxes` can each return a non-`boost::none` `adjustments (account, issuer, currency)`.  What kind of confidence do we have that all of them will return the same `adj->origBalance `?  If they can return different `adj->origBalance` values, how will that play in this code?

seelabs replied:
The `origBalance` is the balance from the ledger before any debits/credits to that trust line _on that payment sandbox_. So `orgBalance` in face will be different on each level on the stack of sandboxes.

You ask how it pays into the code. We take the minimum of these values as a lower bound for our return value. The reason this is needed is:

1) Sometimes the original balance - all the debits is slightly larger than it should be due to loss of precision in the accumulated debits.
2) We already use the current balance as a lower bound, but the account may have received a credit, so that lower bound is too large.

With this scheme, when a sandbox is thrown away, the available balance will never go down (as required by the payment algorithm).

scottschurr replied:
Thanks for the explanation.  I see it now.  The `minBal` is the smallest starting balance seen for this `(account, issuer, currency)` in any of the recursive `PaymentSandboxes`.  So you're finding the smallest post-computation balance regardless of the depth of PaymentSandboxes.  That looks right.

You might consider adding your explanation as a comment in the code.  Personally, I didn't know enough about `PaymentSandbox` internals to get what you were up to just by reading the code.

scottschurr commented:
I left one question, but the change looks reasonable to me.  I cherry picked your commit onto my offer crossing branch and everything looks like it works well there as well.

scottschurr commented:
:+1: Looks reasonable to me.

vinniefalco commented on src/test/app/Flow_test.cpp:
```diff
@@ -1394,6 +1452,7 @@ struct Flow_test : public beast::unit_test::suite
         testSelfFundedXRPEndpoint(true);
         testUnfundedOffer(true);
         testUnfundedOffer(false);
+        testDirectStepReex();
```
Consider `testReExecuteDirectStep`

vinniefalco commented on src/ripple/ledger/impl/PaymentSandbox.cpp:
```diff
@@ -186,15 +186,26 @@ PaymentSandbox::balanceHook (AccountID const& account,
             }
         }
         adjustedAmt = std::min(amount, lastBal - delta);
+        if (amendmentRIPD1368(info().parentCloseTime))
```
Pardon my ignorance but this is not an amendment then? Its a ledger switch?

seelabs replied:
I rebased this on my "Add the config preset features to the view" PR and I'm putting it on a feature now.

vinniefalco commented on src/test/app/Flow_test.cpp:
```diff
@@ -1379,6 +1379,64 @@ struct Flow_test : public beast::unit_test::suite
+
+        // Need to be past this time to see the bug
+        env.close(amendmentRIPD1274SoTime() +
+            100 * env.closed()->info().closeTimeResolution);
```
KInd of annoying that we have to always account for the close time resolution in code like this

seelabs commented:
I just (force) pushed three changes:
1) I rebased onto PR 1981 (Add the config preset features to the view)
2) I renamed one of the test functions (trivial)
3) I replaced the time based switch with an amendment

Only the top two commits need to be reviewed

scottschurr commented:
:+1: Looks swell to me.

seelabs commented:
In 0.60.0-b7
